Add docker-compose based test fixture for Azure

tlrx opened the pull request:
This pull request adds a new `:test:fixtures:azure-fixture` project which provides a docker-compose based container that runs a `AzureHttpFixture` Java class that emulates an Azure Storage service. 

The logic to emulate the service is extracted from existing tests and placed in `AzureHttpHandler` into the new project so that it can be easily reused. The `:plugins:repository-azure` project is an example of such utilization. 

The `AzureHttpFixture` fixture is just a wrapper around `AzureHttpHandler` and is now executed within the docker container. The `:plugins:repository-azure:qa:microsoft-azure` project uses the new test fixture and the existing `AzureStorageFixture` has been removed.

tlrx committed:
azure storage test fixture

elasticmachine commented:
Pinging @elastic/es-core-infra (:Core/Infra/Build)

elasticmachine commented:
Pinging @elastic/es-distributed (:Distributed/Snapshot/Restore)

tlrx commented on plugins/repository-azure/qa/microsoft-azure-storage/build.gradle:
```diff
@@ -38,20 +49,12 @@ String azureSasToken = System.getenv("azure_storage_sas_token")
     azureKey = 'YXp1cmVfaW50ZWdyYXRpb25fdGVzdF9rZXk=' // The key is "azure_integration_test_key" encoded using base64
-    azureContainer = 'container_test'
-    azureBasePath = 'integration_test'
+    azureContainer = 'container'
```
This is changed to accommodate the default values used by the new test fixture.

tlrx committed:
Remove log trace

tlrx commented on test/fixtures/azure-fixture/src/main/java/fixture/azure/AzureHttpHandler.java:
```diff
@@ -0,0 +1,225 @@
+ * Minimal HTTP handler that acts as an Azure compliant server
+ */
+@SuppressForbidden(reason = "Uses a HttpServer to emulate an Azure endpoint")
+public class AzureHttpHandler implements HttpHandler {
```
The implementation is the same as InternalHttpHandler with the following changes:
- container name is passed as a ctor parameter
- some links to the Azure REST APIs are added as comments
- Put Blob now also checks "If-None-Match" header as the AzureStorageFixture did

tlrx commented:
The main motivations around this change are to allow more parallelization when executing tests on CI and to allow new features to easily use Azure fixtures in tests.

original-brownbear reviewed (commented):
This looks good to me, but I'd leave it to @atorok to give his final approval here build wise :)

tlrx committed:
add missing dependsOn

mark-vieira commented on test/fixtures/azure-fixture/build.gradle:
```diff
@@ -0,0 +1,44 @@
+    doLast {
+        file("${testFixturesDir}/shared").mkdirs()
+        project.copy {
+            from project.tasks.fixtureJar.outputs.files
```
This can be simplified to `from jar`.

mark-vieira commented on test/fixtures/azure-fixture/build.gradle:
```diff
@@ -0,0 +1,44 @@
+        file("${testFixturesDir}/shared").mkdirs()
+        project.copy {
+            from project.tasks.fixtureJar.outputs.files
+            from project.configurations.compile
```
This should technically be `from configurations.runtimeClasspath` since that's what we want. It's just in this particular case the compile and runtime classpaths are effectively the same.

mark-vieira reviewed (approved):
Just a couple of comments, otherwise this LGTM.

Thanks for taking care of this. I'm _very_ happy to see usages of `AntFixture` removed. We should definitely do the same for all the others.

atorok commented on plugins/repository-azure/build.gradle:
```diff
@@ -27,6 +27,7 @@ dependencies {
   compile 'com.microsoft.azure:azure-keyvault-core:1.0.0'
   compile 'com.google.guava:guava:20.0'
   compile 'org.apache.commons:commons-lang3:3.4'
+  testCompile project(':test:fixtures:azure-fixture')
```
It would be better not to have multiple types of dependencies  on a project it makes things harder to change, we already depend on it as a fixture it would be nice to avoid depending on the jar as well. 
From what I see it's because the tests also use it to start up some fixtures, could we change that so they talk to one externally ? It would be even better if we could verify the retries as a unit tests without  talking to the actual service, e.x. by mocking the sdk ? I understand if that's more work that we don't want to take on right now, but it would be a nice to have.

tlrx replied:
> It would be better not to have multiple types of dependencies on a project it makes things harder to change, we already depend on it as a fixture it would be nice to avoid depending on the jar as well.

Yes, I agree. I wanted to have the fixture plumbing and its logic located at the same place within the same project and this is the only way I know to do it. The `AzureHttpHandler` could be placed into the test framework (and the fixture depends on it) but I don't think this a better solution.

> From what I see it's because the tests also use it to start up some fixtures, could we change that so they talk to one externally ?

For `AzureBlobStoreRepositoryTests` having the server logic running in the same JVM is really convenient when it comes to understand or debug things. I guess we could change this to run on external servers but we'll lost the ability to debug easily.

> It would be even better if we could verify the retries as a unit tests without talking to the actual service, e.x. by mocking the sdk ?

This is what was done for a long time (and still done in some tests) but for `ESBlobStoreRepositoryIntegTestCase` and other `*RetriesTests` we want to test how the SDK behaves exactly in various scenarios of server side errors. The SDKs offer some way to mock the HTTP calls (either using Mockito or some kind of low-level execution mock) but it does not reflect the exact behavior of the SDK and issues like #46589 would have been hard to troubleshoot and fix.

tlrx replied:
> I guess we could change this to run on external servers but we'll lost the ability to debug easily.

Thinking about this twice, we could maybe just always run the fixture within the container in debug mode and connect the IDE to the ephemeral port

atorok replied:
I'll leave this to your best judgment. I don't feel strongly about not having this dependency and you bring up good arguments for not doing it too.

atorok commented on test/fixtures/azure-fixture/build.gradle:
```diff
@@ -0,0 +1,44 @@
+    dependsOn fixtureJar
+    doLast {
+        file("${testFixturesDir}/shared").mkdirs()
+        project.copy {
```
The jar task has a convention for it's output, we could avoid the copy here and hard code that into `docker-compose`

tlrx replied:
> and hard code that into `docker-compose`

You mean, exposing `./build/distributions` as a `volume` in `docker-compose.yml`? That should work, but what about dependencies?

atorok replied:
We do need to copy those to get them into the container,  copying the jar is actually better, I missed the dependencies so this is fine.

alpar-t reviewed (commented):
Do we have a compelling reason to roll out own ? It seems there is an implementation of azure called [azurite](https://github.com/azure/azurite) pointed to in the [azure docs](https://docs.microsoft.com/en-us/azure/storage/common/storage-use-emulator).

original-brownbear commented:
@atorok last time I tried (little less than a year ago) Azurite wasn't compatible enough to work for us. We also recently tried using it in a Cloud project and couldn't get it to work with SDK v8 (which we're using in ES and newer SDKs are unstable sadly :(). We should keep an eye on it though, maybe we can make use of it eventually. It seems they are making progress in that project :)

tlrx committed:
Mark's feedback

tlrx commented:
> Do we have a compelling reason to roll out own ? It seems there is an implementation of azure called [azurite](https://github.com/azure/azurite) pointed to in the [azure docs](https://docs.microsoft.com/en-us/azure/storage/common/storage-use-emulator).

I think Armin already mentioned the reasons to not use azurite (yet) but it seems that they are now almost feature complete (Blob APIs) so I hope we'll use it in the future.

tlrx commented:
Thanks @mark-vieira @original-brownbear and @atorok for your feedback. I addressed most of the comments or I asked precision when needed. Please let me know what you think!

alpar-t reviewed (approved):
LGTM
